File: implementations/java/server/src/test/java/com/theshuai/specusserver/http/DecompressionLimitsTests.java
package com.theshuai.specusserver.http;

import org.junit.jupiter.api.Test;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;

class DecompressionLimitsTests {
    @Test
    void ordinaryBodiesDecompressUnchanged() throws Exception {
        byte[] payload = "the quick brown fox. ".repeat(2000).getBytes();
        byte[] compressed = gzip(payload);

        try (GZIPInputStream input = new GZIPInputStream(new ByteArrayInputStream(compressed))) {
            assertArrayEquals(payload,
                    DecompressionLimits.readAllBounded(input, compressed.length));
        }
    }

    /** A few kilobytes of crafted gzip expands to gigabytes. Reading to EOF would end the process. */
    @Test
    void decompressionBombIsRefused() throws Exception {
        byte[] bomb = gzip(new byte[32 * 1024 * 1024]);

        try (GZIPInputStream input = new GZIPInputStream(new ByteArrayInputStream(bomb))) {
            assertThrows(DecompressionLimits.LimitExceededException.class,
                    () -> DecompressionLimits.readAllBounded(input, bomb.length));
        }
    }

    @Test
    void limitCombinesTheAbsoluteAndRatioCaps() {
        // Tiny inputs get the flat allowance, so framing overhead cannot make the ratio meaningless.
        assertEquals(DecompressionLimits.MIN_RATIO_ALLOWANCE_BYTES, DecompressionLimits.limitFor(0));
        assertEquals(DecompressionLimits.MIN_RATIO_ALLOWANCE_BYTES, DecompressionLimits.limitFor(-1));

        // In the middle the ratio binds.
        assertEquals(256 * 1024 * DecompressionLimits.MAX_RATIO,
                DecompressionLimits.limitFor(256 * 1024));

        // Past that the absolute cap binds, including for a size large enough to overflow a naive
        // multiplication.
        assertEquals(DecompressionLimits.MAX_DECOMPRESSED_BYTES,
                DecompressionLimits.limitFor(1024 * 1024));
        assertEquals(DecompressionLimits.MAX_DECOMPRESSED_BYTES,
                DecompressionLimits.limitFor(Integer.MAX_VALUE));
    }

    /** A body that exactly fills its allowance is legitimate and must not be rejected. */
    @Test
    void bodyWithinTheAllowanceIsAccepted() throws Exception {
        byte[] payload = new byte[DecompressionLimits.MIN_RATIO_ALLOWANCE_BYTES];
        java.util.Arrays.fill(payload, (byte) 'x');
        byte[] compressed = gzip(payload);

        try (GZIPInputStream input = new GZIPInputStream(new ByteArrayInputStream(compressed))) {
            assertEquals(payload.length,
                    DecompressionLimits.readAllBounded(input, compressed.length).length);
        }
    }

    private static byte[] gzip(byte[] payload) throws Exception {
        ByteArrayOutputStream buffer = new ByteArrayOutputStream();
        try (GZIPOutputStream compressor = new GZIPOutputStream(buffer)) {
            compressor.write(payload);
        }
        return buffer.toByteArray();
    }
}

File: implementations/csharp/server/tests/Specus.IntegrationTests/DecompressionLimitsTests.cs
using System.IO.Compression;
using System.Text;
using Specus.Server.Http;

namespace Specus.IntegrationTests;

public sealed class DecompressionLimitsTests
{
    [Fact]
    public void OrdinaryBodiesDecompressUnchanged()
    {
        var payload = Encoding.UTF8.GetBytes(string.Concat(Enumerable.Repeat("the quick brown fox. ", 2000)));
        var compressed = Gzip(payload);

        using var source = new MemoryStream(compressed);
        using var decoder = new GZipStream(source, CompressionMode.Decompress);
        Assert.Equal(payload, DecompressionLimits.ReadAllBounded(decoder, compressed.Length));
    }

    /// <summary>A few kilobytes of crafted gzip expands to gigabytes; reading to the end would end the process.</summary>
    [Fact]
    public void DecompressionBombIsRefused()
    {
        var bomb = Gzip(new byte[32 * 1024 * 1024]);

        using var source = new MemoryStream(bomb);
        using var decoder = new GZipStream(source, CompressionMode.Decompress);
        Assert.Throws<DecompressionLimits.LimitExceededException>(
            () => DecompressionLimits.ReadAllBounded(decoder, bomb.Length));
    }

    [Fact]
    public void LimitCombinesTheAbsoluteAndRatioCaps()
    {
        // Tiny inputs get the flat allowance, so framing overhead cannot make the ratio meaningless.
        Assert.Equal(DecompressionLimits.MinRatioAllowanceBytes, DecompressionLimits.LimitFor(0));
        Assert.Equal(DecompressionLimits.MinRatioAllowanceBytes, DecompressionLimits.LimitFor(-1));

        // In the middle the ratio binds.
        Assert.Equal(256 * 1024 * DecompressionLimits.MaxRatio,
            DecompressionLimits.LimitFor(256 * 1024));

        // Past that the absolute cap binds, including for a size large enough to overflow a naive
        // multiplication.
        Assert.Equal(DecompressionLimits.MaxDecompressedBytes,
            DecompressionLimits.LimitFor(1024 * 1024));
        Assert.Equal(DecompressionLimits.MaxDecompressedBytes,
            DecompressionLimits.LimitFor(int.MaxValue));
    }

    /// <summary>A body that exactly fills its allowance is legitimate and must not be rejected.</summary>
    [Fact]
    public void BodyWithinTheAllowanceIsAccepted()
    {
        var payload = new byte[DecompressionLimits.MinRatioAllowanceBytes];
        Array.Fill(payload, (byte)'x');
        var compressed = Gzip(payload);

        using var source = new MemoryStream(compressed);
        using var decoder = new GZipStream(source, CompressionMode.Decompress);
        Assert.Equal(payload.Length,
            DecompressionLimits.ReadAllBounded(decoder, compressed.Length).Length);
    }

    private static byte[] Gzip(byte[] payload)
    {
        using var buffer = new MemoryStream();
        using (var compressor = new GZipStream(buffer, CompressionMode.Compress, leaveOpen: true))
        {
            compressor.Write(payload, 0, payload.Length);
        }
        return buffer.ToArray();
    }
}
